src/persist/src/indexed/mod.rs

```diff
@@ -537,35 +553,62 @@ impl<L: Log, B: Blob> Indexed<L, B> {
         Ok(())
     }
 
-    /// Validates the following postconditions for draining pending requests:
-    ///
-    /// - The meta we might roll back to must be equal to the durably
-    ///   persisted meta.
-    /// - There are no more pending responses or pending writes.
-    #[cfg(any(debug_assertions, test))]
-    fn validate_drain_pending_postconditions(&self) -> Result<(), Error> {
-        // The postconditions are strictly more general than the preconditions so validate those as well.
-        self.validate_drain_pending_preconditions()?;
-
-        self.pending.validate_empty()?;
+    fn validate_pending_empty(&self) -> Result<(), Error> {
+        if let Some(pending) = self.pending.as_ref() {
+            return Err(Error::from(format!(
+                "still have pending, expected None: {:?}",
+                pending
+            )));
+        }
         Ok(())
     }
 
+    /// Return true if Pending has at least one pending response.
+    pub fn has_pending_responses(&self) -> bool {
+        self.pending
+            .as_ref()
+            .map_or(false, |p| !p.responses.is_empty())
+    }
+
     /// Commit any pending in-memory changes to persistent storage, respond to clients
     /// and notify any listeners.
     fn drain_pending(&mut self) -> Result<(), Error> {
-        #[cfg(any(debug_assertions, test))]
-        {
-            assert_eq!(self.validate_drain_pending_preconditions(), Ok(()));
-        }
-        let ret = match self.drain_pending_inner() {
-            Ok(_) => {
-                let mut responses = self.pending.take_responses();
+        debug_assert_eq!(self.validate(), Ok(()));
+
+        let pending = match self.pending.take() {
+            Some(pending) => pending,
+            None => return Ok(()),
+        };
+
+        let meta_before = pending.durable_meta;
+        let updates_by_id = pending.writes;
+        let seals_for_listeners = pending.seals;
+        let updates_for_listeners = updates_by_id.clone();
+
+        let ret = (|| {
+            // TODO: The following error handling took a while to debug, see if
```

Comment:
I've been staring at this for a while and I think _maybe_ if we could make `drain_pending_writes` be called with `apply_unbatched_command` that might be a nice simplification? It might not for a reason I don't see just yet

Reply:
Yeah it seems like that should work to me too, but it demonstrably doesn't (our existing tests fail). I thought about this yesterday for 10min and couldn't figure out what was wrong with using apply_unbatched_command for this and I did it again today and still can't (and confirmed that it still does break the tests). My takeaway is that all this stuff is subtle enough that I don't want to change it at the last minute of this PR, so I'll merge this with the TODO, but I do think it's worth getting to the bottom of this.